storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -2043,8 +2058,7 @@ fseg_alloc_free_page_low(
 		goto take_hinted_page;
 		/*-----------------------------------------------------------*/
 	} else if ((direction != FSP_NO_DIR)
-		   && ((reserved - used) < reserved / FSEG_FILLFACTOR)
-		   && (used >= FSEG_FRAG_LIMIT)
+		   && need_for_new_extent(used, reserved)
 		   && (ret_descr = fseg_alloc_free_extent(seg_inode, iblock,
 							  &xdes, space,
 							  mtr, err))) {
```

Comment:
This is very similar to the previous `else if` condition. I think that we should refactor this so that the condition `need_for_new_extent()`  needs to be evaluated only once. Here is one idea:
```c++
	} else if (need_for_new_extent(used, reserved)) {
		goto no_new_extent;
	} else if (xdes_get_state(descr) == XDES_FREE) {
		// …
	} else if (direction != FSP_NO_DIR
		   && (ret_descr = fseg_alloc_free_extent(seg_inode, iblock,
							  &xdes, space,
							  mtr, err))) {
		// …
	} else if (UNIV_UNLIKELY(*err != DB_SUCCESS)) {
		return nullptr;
	} else {
	no_new_extent:
		// … the rest of the "else if" chain, the first one being an "if"
	}
```
Even better would be to rewrite part of the `else if` chain as `switch (xdes_get_state(descr))`. While doing that, I would reformat the function away from the archaic InnoDB style.